lib/mode.coffee

```diff
@@ -0,0 +1,244 @@
+  Promise.resolve().then () =>
+    rustcArgs = switch linter.rustcBuildTest
+      when true then ['--cfg', 'test', '-Z', 'no-trans', '--color', 'never']
+      else ['-Z', 'no-trans', '--color', 'never']
+    cmd = [linter.rustcPath]
+      .concat rustcArgs
+    if cargoManifestPath
+      cmd.push '-L'
+      cmd.push path.join path.dirname(cargoManifestPath), linter.cargoDependencyDir
+    compilationFeatures = linter.compilationFeatures(false)
+    cmd = cmd.concat compilationFeatures if compilationFeatures
+    cmd = cmd.concat [editingFile]
+    [editingFile, cmd]
+
+cachedUsingMultitoolForClippy = null
+
+buildCargoArguments = (linter, cargoManifestPath) ->
+  buildCargoPath = (cargoPath, cargoCommand) ->
+    # the result is cached to avoid delays
+    if cachedUsingMultitoolForClippy? and do cachedUsingMultitoolForClippy.valid
+      Promise.resolve().then () =>
+        do cachedUsingMultitoolForClippy.getResult
+    else
+      # Decide if should use older multirust or newer rustup
+      usingMultitoolForClippy =
+        atom_linter.exec 'rustup', ['--version'], {ignoreExitCode: true}
+          .then ->
+            result: true, tool: 'rustup'
+          .catch ->
+            # Try to use older multirust at least
+            atom_linter.exec 'multirust', ['--version'], {ignoreExitCode: true}
+              .then ->
+                result: true, tool: 'multirust'
+              .catch ->
+                result: false
+      usingMultitoolForClippy.then (canUseMultirust) ->
+        if cargoCommand == 'clippy' and canUseMultirust.result
+          [canUseMultirust.tool, 'run', 'nightly', 'cargo']
+        else
+          [cargoPath]
+      .then (cached) =>
+        cachedUsingMultitoolForClippy = new CachedResult(cached, 20)
+        cached
+
+  cargoArgs = switch linter.cargoCommand
+    when 'check' then ['check']
+    when 'test' then ['test', '--no-run']
+    when 'rustc' then ['rustc', '-Zno-trans', '--color', 'never']
+    when 'clippy' then ['clippy']
+    else ['build']
+
+  compilationFeatures = linter.compilationFeatures(true)
+  buildCargoPath(linter.cargoPath, linter.cargoCommand).then (cmd) ->
+    cmd = cmd
+      .concat cargoArgs
+      .concat ['-j', linter.jobsNumber]
+    cmd = cmd.concat compilationFeatures if compilationFeatures
+    cmd = cmd.concat ['--manifest-path', cargoManifestPath]
+    [cargoManifestPath, cmd]
+
+# These define the behabiour of each error mode linter-rust has
+errorModes =
+  JSON_RUSTC:
+    neededOutput: (stdout, stderr) ->
+      stderr
+
+    parse: (output, options) =>
+      parseJsonOutput output, options
+
+    buildArguments: (linter, file) ->
+      buildRustcArguments(linter, file).then (cmd_res) ->
+        [file, cmd] = cmd_res
+        cmd = cmd.concat ['--error-format=json']
+        [file, cmd]
+
+  JSON_CARGO:
+    neededOutput: (stdout, stderr) ->
+      stdout
+
+    parse: (output, options) ->
+      options.additionalFilter = (json) ->
```

Comment:
I think this is a bug, on my machine it was filtering out 100% of errors returned from the compiler. I think that either "json" here should be "input" to match below, or vice versa.